File: cpp/app/algorithm_comparison.cpp
/**
 * @file algorithm_comparison.cpp
 * @author Carl Stahoviak
 * @brief A comparison of the planning and optimization algorithm currently
 * implemented.
 * @version 0.1
 * @date 2024-10-01
 * 
 * TODO: Valgrind provides the following summary of memory leaks. Only the
 * "Naive" planner has a leak.
 * 
 * ==37798== LEAK SUMMARY:
 * ==37798==    definitely lost: 128 bytes in 1 blocks
 * ==37798==    indirectly lost: 738 bytes in 6 blocks
 * ==37798==      possibly lost: 0 bytes in 0 blocks
 * ==37798==    still reachable: 0 bytes in 0 blocks
 * ==37798==         suppressed: 0 bytes in 0 blocks
 * 
 * @copyright Copyright (c) 2024
 */
#include "supercharger/algorithm/dijkstras.h"
#include "supercharger/logging.h"
#include "supercharger/supercharger.h"

// #include <glog/logging.h>

#include <algorithm>
#include <iomanip>
#include <iostream>
#include <map>

using namespace supercharger;

template<typename A, typename B>
std::pair<B, A> flip_pair(const std::pair<A, B>& p) {
    return std::pair<B,A>(p.second, p.first);
}

template<typename A, typename B>
std::multimap<B, A, std::greater<B>> flip_map(const std::map<A, B>& src) {
    std::multimap<B, A, std::greater<B>> dst;
    std::transform(
      src.begin(), src.end(), std::inserter(dst, dst.begin()), flip_pair<A, B>);
    return dst;
}

int main(int argc, char** argv)
{
  if ( argc != 3 ) {
    std::cout << "Error: requires initial and final supercharger names."
      << std::endl;        
    return -1;
  }

  // TODO: Add glog
  // google::InitGoogleLogging(argv[0]);
  
  // Parse the initial and goal charger names
  const std::string initial_charger_name = argv[1];
  const std::string goal_charger_name = argv[2];

  // Store a map of planning algorithms and route costs
  std::map<std::string, double> algorithms;

  // Create the Route Planner
  Supercharger app(NaiveCostFcnType::MINIMIZE_TIME_REMAINING);

  // Set the vehicle's speed and max range
  app.max_range() = 320;
  app.speed() = 105;

  // Test to make sure the network getter works.
  const auto& network = app.network();

  // Plan the route via the "Naive" route planner
  PlannerResult result1 = app.PlanRoute(
    initial_charger_name,
    goal_charger_name
  );
  algorithms["NAIVE\t\t\t\t"] = result1.cost;

  // Plan the route via Dijkstra's algorithm.
  app.SetPlanner(AlgoType::DIJKSTRAS, algorithm::SimpleCost);
  PlannerResult result2 = app.PlanRoute(
    initial_charger_name,
    goal_charger_name
  );
  algorithms["DIJKTRA'S\t\t\t"] = result2.cost;

  // Improve on Dijkstra's via the "naive" optimizer.
  app.SetOptimizer(OptimizerType::NAIVE);
  PlannerResult result3 = app.PlanRoute(
    initial_charger_name,
    goal_charger_name
  );
  algorithms["DIJKTRA'S + NAIVE OPTIMIZER\t"] = result3.cost;

  // Improve on Dijkstra's via the "NLOPT" optimizer.
  app.SetOptimizer(OptimizerType::NLOPT);
  PlannerResult result4 = app.PlanRoute(
    initial_charger_name,
    goal_charger_name
  );
  algorithms["DIJKTRA'S + NLOPT\t\t"] = result4.cost;

  // Now sorted by cost.
  std::multimap<double, std::string, std::greater<double>> algs_by_cost = 
    flip_map(algorithms);
    
  // Output the results
  std::cout << "\nPlanning Algorithm\t\tCost [hrs]" << std::endl;
  for ( auto iter = algs_by_cost.cbegin(); iter != algs_by_cost.cend(); ++iter) {
    std::cout << iter->second << std::setprecision(6) << iter->first
      << std::endl;
  }
  std::cout << "\n";
}

File: python/example/algorithm_comparison.py
"""
Recreates the cpp/app/algorithm_comparison.cpp" app on the python side.
"""
from collections import OrderedDict, namedtuple
from time import perf_counter

from supercharger.optimize import optimized_cost

from pysupercharger import (
    dijkstras_simple_cost,
    NaiveCostType,
    OptimizerType,
    Supercharger
)

AlgoStats = namedtuple('AlgoStats', ['time', 'cost'])


if __name__ == "__main__":
    # Define the route's endpoints
    origin = "Council_Bluffs_IA"
    destination = "Cadillac_MI"

    # These endpoints will show a cost difference between the final two algos.
    # origin = "Burlington_WA"
    # destination = "Marathon_FL"

    # Store a mapping of algorithms and route costs
    algorithms = OrderedDict()

    # Create the supercharger using Dijkstra's algorithm
    supercharger = Supercharger(
        naive_cost_type=NaiveCostType.MINIMIZE_TIME_REMAINING)

    # Set the vehicle's speed and max range
    supercharger.max_range = 320
    supercharger.speed = 105

    # Plan the route via the "Naive" route planner.
    start = perf_counter()
    result1 = supercharger.plan_route(origin, destination)
    stop = perf_counter()
    algorithms["Naive\t\t\t\t\t\t"] = \
        AlgoStats(time=stop - start, cost=result1.cost)

    # Plan the route via Dijkstra's algorithm.
    supercharger.set_planner(cost_f=dijkstras_simple_cost)
    start = perf_counter()
    result2 = supercharger.plan_route(origin, destination)
    stop = perf_counter()
    algorithms["Dijkstra's\t\t\t\t\t"] = \
        AlgoStats(time=stop - start, cost=result2.cost)

    # Improve on Dijkstra's via the "NLOPT" optimizer.
    supercharger.set_optimizer(type=OptimizerType.NLOPT)
    start = perf_counter()
    result3 = supercharger.plan_route(origin, destination)
    stop = perf_counter()
    algorithms["Dijkstra's + Post Optimization"] = \
        AlgoStats(time=stop - start, cost=result3.cost)

    # Use Dijkstra's algorithm with an optimized cost function
    supercharger.set_planner(cost_f=optimized_cost)
    supercharger.set_optimizer(type=OptimizerType.NONE)
    start = perf_counter()
    result4 = supercharger.plan_route(origin, destination)
    stop = perf_counter()
    algorithms["Dijkstra's + Optimized Cost\t"] = \
        AlgoStats(time=stop - start, cost=result4.cost)

    print("\nPlanning Algorithm\t\t\t\tTime [ms]\tCost [hrs]")
    for algo_name, stats in algorithms.items():
        print(f'{algo_name}\t{stats.time * 1e3:.2f}\t\t{stats.cost:.4f}')
